### ISSUE TITLE
Renderer flags components as disposed even though they are not (for IAsyncDisposable) in Blazor

### ISSUE THREAD

**Issue by linkdotnet** [dotnet/aspnetcore#41017]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

If we are handling with an `IAsyncDisposable` the Blazor `Renderer` flags an `IAsyncDisposable` as  disposed as soon as it exists the `await` boundary of the  `DisposeAsync` method. Imagine the following simplistic component:

Sample-Component with the component Id 2:
```csharp
private sealed class AsyncDisposableComponent : ComponentBase, IAsyncDisposable
{
    public bool WasDisposed { get; private set; }
    public async ValueTask DisposeAsync()
    {
        // After the await the Renderer flags the component as disposed
	await Task.Delay(50000);
	
        // Huh? But we still have work to do!
        WasDisposed = true;
    }
}
```

The observed behavior is, that after `await Task.Delay(50000);` the component id is put into the `_batchBuilder.DisposedComponentIds` collection. Arguable the component is currently disposing but it is not done and therefore not disposed.

The issue comes from the [renderer](https://source.dot.net/#Microsoft.AspNetCore.Components/RenderTree/Renderer.cs,800) itself. In the case of an `IAsyncDisposable` if the `Task` is completed everything fine. If not, and this is our case, the Renderer just puts it into a queue where it gets handled via `[AddToPendingTasks](https://source.dot.net/#Microsoft.AspNetCore.Components/RenderTree/Renderer.cs,830)` which indeed awaits internally but does not expose the task for the renderer to be awaited.

So the Renderer directly drops out of the function and puts the component id into the array via [`_batchBuilder.DisposedComponentIds.Append(disposeComponentId);`](https://source.dot.net/#Microsoft.AspNetCore.Components/RenderTree/Renderer.cs,848)

### Expected Behavior

The component id is only added to the collection when the underlying dispose `Task` is really done. 

### Steps To Reproduce

See description for reproducing the issue

### .NET Version

6.0.201

### Anything else?

Output of `dotnet --info`:

```
$ dotnet --info
.NET SDK (reflecting any global.json):
 Version:   6.0.201
 Commit:    ef40e6aa06

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19043
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.201\

Host (useful for support):
  Version: 6.0.3
  Commit:  c24d9a9c91

.NET SDKs installed:
  3.0.100 [C:\Program Files\dotnet\sdk]
  3.1.100 [C:\Program Files\dotnet\sdk]
  5.0.100 [C:\Program Files\dotnet\sdk]
  5.0.102 [C:\Program Files\dotnet\sdk]
  5.0.406 [C:\Program Files\dotnet\sdk]
  6.0.100-rc.2.21505.57 [C:\Program Files\dotnet\sdk]
  6.0.101 [C:\Program Files\dotnet\sdk]
  6.0.102 [C:\Program Files\dotnet\sdk]
  6.0.201 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.22 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.23 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.0-rc.2.21480.10 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.22 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.23 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.0-rc.2.20475.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.15 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0-rc.2.21480.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.22 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.23 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.0-rc.2.20475.6 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.15 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.0-rc.2.21501.6 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.3 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

**Comment by javiercn** [comment]
@linkdotnet thanks for contacting us.

The component is disposed as part of the batch, whether it does synchronously or asynchronously doesn't affect that fact. Is there a concrete scenario where you are having an issue with this behavior?

**Comment by linkdotnet** [comment]
Hey @javiercn thanks for the swift response.

Yes I have a concrete use case, even though it's very special. Anyway: for [bUnit](https://github.com/bUnit-dev/bUnit/) (unit test library for Blazor) we introduced a feature where we basically detach the component under test from the render tree, so that the component (plus its children) get disposed. For a synchronous `IDisposable` that is no issue, as we can just wait for the `Task` to complete. But this falls short for `IAsyncDisposable` as the renderer directly adds the component as disposed as soon as we hit the first `await` inside `DisposeAsync`.

Basically [this](https://github.com/bUnit-dev/bUnit/blob/main/src/bunit.core/Rendering/TestRenderer.cs#L137) line will evaluate to `true` even though `DisposeAsync` is not finished.

From a Blazor point of view I can understand that as the component never ever again will invoke another render cycle. Just wanted to clarify if this is intended behavior.

**Comment by javiercn** [comment]
@linkdotnet thanks for the additional details.

Unfortunately, I'm not sure if this is something we can easily change or want to. The way we think about it is that as part of the render process the components get disposed, and that can happen synchronously or asynchronously. However, we don't want to wait for a component to be disposed to inform/finish processing the render batch for several reasons:
* It'll slow down rendering.
* It'll make rendering asynchronous.

We don't care about when the disposal process finishes as long as it does, and we get to capture and surface any potential error. The `AddToPendingTasks` is to make sure we can reach `quiescence` during prerendering as otherwise we wouldn't be able to surface disposal errors.

Hope this helps clarify the design decisions around this area. You might need to find an alternative way to achieve your goal here, since this behavior is deliberate.
